actors/miner/src/lib.rs

```diff
@@ -3694,38 +3718,111 @@ fn extend_sector_committment(
     curr_epoch: ChainEpoch,
     reward_stats: &ThisEpochRewardReturn,
     power_stats: &ext::power::CurrentTotalPowerReturn,
+    circulating_supply: &TokenAmount,
     new_expiration: ChainEpoch,
     sector: &SectorOnChainInfo,
     sector_size: SectorSize,
     claim_space_by_sector: &BTreeMap<SectorNumber, (u64, u64)>,
 ) -> Result<SectorOnChainInfo, ActorError> {
     validate_extended_expiration(policy, curr_epoch, new_expiration, sector)?;
+    let mut new_sector = sector.clone();
+    new_sector.expiration = new_expiration;
+    new_sector.power_base_epoch = curr_epoch;
 
-    // all simple_qa_power sectors with VerifiedDealWeight > 0 MUST check all claims
-    if sector.simple_qa_power {
-        extend_simple_qap_sector(
-            policy,
-            new_expiration,
-            curr_epoch,
-            reward_stats,
-            power_stats,
-            sector,
-            sector_size,
-            claim_space_by_sector,
-        )
-    } else {
-        extend_non_simple_qap_sector(new_expiration, curr_epoch, sector)
+    let new_duration = new_sector.expiration - new_sector.power_base_epoch;
+    let old_duration = sector.expiration - sector.power_base_epoch;
+    let remaining_time = sector.expiration - curr_epoch;
+
+    if sector.deal_weight.is_positive() {
+        new_sector.deal_weight = (&sector.deal_weight * remaining_time) / old_duration;
     }
+
+    if sector.verified_deal_weight.is_positive() {
+        if !sector.simple_qa_power {
+            // divide by the old diration to get the verified deal size.
+            // multiply by the remaining time to calculate the remaining weight. This remaining
+            // weight will end up getting spread out over the _new_ duration.
+            new_sector.verified_deal_weight =
+                (&sector.verified_deal_weight * remaining_time) / old_duration;
+        } else {
+            let old_verified_deal_space = &sector.verified_deal_weight / old_duration;
```

Comment:
This may lose precision (divide then multiply). Is that ok?

Reply:
I don't think it will because old_duration should always be a factor of verified_deal_weight by construction

Reply:
Hm. Yes, you should be right. I'll update that.

Reply:
Actually. Yes, in this case, no in the other. I'll move the conversation back there.

Reply:
Nvm, let's just discuss it here. In the case of non-simple QaP, we multiply by "remaining time" not the "new duration". So I think we would lose precision.

Reply:
Ok yeah I agree this isn't lossless in the non-simple case on the first extension because we divide by the whole duration and legacy weights are not a multiple of whole duration.  After the first extension then there is no more precision lost since weight is always a multiple of the old_duration as enforce here.

Even if we don't explicitly divide out by time to get deal space first I think the intent to document is what you have written here.